### PR TITLE
Add initial lowering of a trivial function

### PR DESCRIPTION
**Pull request by jonmeow**
This is starting to build out actual lowering logic, for a really simple `fn Main() -> i32 { return 0; }`

Notes for achieving this:

- In semantics, currently function names are bound separate from the signature. When emitting IR, this turns out to be inconvenient because we want to know the name when we process the declaration and the definition. This change addresses that by merging the name into the FunctionDeclaration node, which is also accessible from the definition. It removes the separate BindName. This should be the cause of all the test changes in semantics, because the IR generated changes.

- Add a "Lowering" class which I'm using to hold the llvm builder state. This class now has minimal support for the SemanticsIR generated by the above example.

- In the "Lowering", values from expressions are stored in a DenseMap. I'll keep thinking about whether there's a cleaner way to achieve this, and I'd call it a temporary solution for now. However, this is how the `0` in `return 0` gets properly associated across SemanticsIR instructions, and it'll frequently be an issue in less trivial cases.

### CODE REVIEW

File: toolchain/driver/driver.cpp [hunk 1]
```diff
@@ -209,6 +209,11 @@ auto Driver::RunDumpSubcommand(DiagnosticConsumer& consumer,
   }
   CARBON_VLOG() << "semantics_ir: " << semantics_ir;
 
+  // Unlike previous steps, errors block further progress.
+  if (has_errors) {
+    return false;
```
Comment (chandlerc) [inline toolchain/driver/driver.cpp]: Maybe at least a `VLOG` here rather than relying on the absence of the one below?

**Review by chandlerc — APPROVED** [review]
LGTM

File: toolchain/driver/driver.cpp (hunk 1 above)
Comment (jonmeow) [reply toolchain/driver/driver.cpp]: Right now, we don't vlog when a step results in an error; are you wanting that to be changed in general? I'd probably do it on line 204 and similar, where the error is returned from a call, not here.

Note though, I've typically been vlogging things that aren't already printed. In the case of errors, there'll be diagnostic output indicating the error, and the tool should return an error [exit code].

Comment (jonmeow) [reply toolchain/driver/driver.cpp]: I'm adding a VLOG per the request.

Comment (chandlerc) [reply toolchain/driver/driver.cpp]: I was specifically hoping for a VLOG to indicate that at this point we have to give up. Technically, the lack of the IR dump vlog tells you that, but it seemed nice to be explicit that this is in fact the fatal point in the process.

Anyways, still LG and thanks!

**Pull request merged**